File: migration-samples/backend/app/api/v1/agents.py
"""
FastAPI agent endpoints
Migrated from Flask agent_routes.py
"""
from fastapi import APIRouter, HTTPException, Depends, BackgroundTasks
from typing import Dict, Any
import asyncio
import time
import logging

from app.models.agent import (
    AgentRunRequest, 
    AgentFollowupRequest, 
    AgentResponse, 
    ModelListResponse,
    AgentStatus
)
from app.services.agent_service import AgentService
from app.core.config import settings

logger = logging.getLogger(__name__)
router = APIRouter()


def get_agent_service() -> AgentService:
    """Dependency to get agent service instance"""
    return AgentService()


@router.post("/agents/run", response_model=AgentResponse)
async def run_agent(
    request: AgentRunRequest,
    agent_service: AgentService = Depends(get_agent_service)
) -> AgentResponse:
    """
    Run an AI agent for data transformation
    Migrated from Flask /api/agent/run endpoint
    """
    start_time = time.time()
    
    try:
        logger.info(f"Running agent {request.agent_type} for session {request.session_id}")
        
        # Run agent in thread pool to avoid blocking
        result = await asyncio.get_event_loop().run_in_executor(
            None,
            agent_service.run_agent,
            request
        )
        
        execution_time = time.time() - start_time
        result.execution_time = execution_time
        
        logger.info(f"Agent {request.agent_type} completed in {execution_time:.2f}s")
        return result
        
    except Exception as e:
        logger.error(f"Agent execution failed: {str(e)}")
        return AgentResponse(
            status=AgentStatus.ERROR,
            agent=request.agent_type.value,
            error_message=str(e),
            execution_time=time.time() - start_time
        )


@router.post("/agents/followup", response_model=AgentResponse)
async def followup_agent(
    request: AgentFollowupRequest,
    agent_service: AgentService = Depends(get_agent_service)
) -> AgentResponse:
    """
    Follow up with an AI agent for iterative refinement
    """
    start_time = time.time()
    
    try:
        logger.info(f"Following up with agent {request.agent_type} for session {request.session_id}")
        
        result = await asyncio.get_event_loop().run_in_executor(
            None,
            agent_service.followup_agent,
            request
        )
        
        execution_time = time.time() - start_time
        result.execution_time = execution_time
        
        return result
        
    except Exception as e:
        logger.error(f"Agent followup failed: {str(e)}")
        return AgentResponse(
            status=AgentStatus.ERROR,
            agent=request.agent_type.value,
            error_message=str(e),
            execution_time=time.time() - start_time
        )


@router.get("/agents/models", response_model=ModelListResponse)
async def get_available_models(
    agent_service: AgentService = Depends(get_agent_service)
) -> ModelListResponse:
    """
    Get list of available LLM models and providers
    Migrated from Flask /api/get-models endpoint
    """
    try:
        models_data = await asyncio.get_event_loop().run_in_executor(
            None,
            agent_service.get_available_models
        )
        
        return ModelListResponse(
            models=models_data.get("models", []),
            providers=models_data.get("providers", [])
        )
        
    except Exception as e:
        logger.error(f"Failed to get available models: {str(e)}")
        raise HTTPException(status_code=500, detail=str(e))


@router.get("/agents/config")
async def get_agent_config() -> Dict[str, Any]:
    """
    Get agent configuration settings
    """
    return {
        "formulate_timeout_seconds": settings.FORMULATE_TIMEOUT_SECONDS,
        "max_repair_attempts": settings.MAX_REPAIR_ATTEMPTS,
        "default_chart_width": settings.DEFAULT_CHART_WIDTH,
        "default_chart_height": settings.DEFAULT_CHART_HEIGHT,
        "exec_python_in_subprocess": settings.EXEC_PYTHON_IN_SUBPROCESS,
        "disable_display_keys": settings.DISABLE_DISPLAY_KEYS
    }


@router.post("/agents/explain-code")
async def explain_code(
    request: Dict[str, Any],
    agent_service: AgentService = Depends(get_agent_service)
) -> AgentResponse:
    """
    Explain generated code in natural language
    """
    try:
        result = await asyncio.get_event_loop().run_in_executor(
            None,
            agent_service.explain_code,
            request.get("code", ""),
            request.get("model_config", {}),
            request.get("session_id", "")
        )
        
        return result
        
    except Exception as e:
        logger.error(f"Code explanation failed: {str(e)}")
        return AgentResponse(
            status=AgentStatus.ERROR,
            agent="code_explanation",
            error_message=str(e)
        )

File: migration-samples/frontend/lib/api/agents.ts
/**
 * Agent API client
 * Handles AI agent-related API calls to FastAPI backend
 */

import { apiClient, APIError } from './client'

export interface ModelConfig {
  provider: string
  model: string
  api_key?: string
  api_base?: string
  api_version?: string
  temperature: number
  max_tokens?: number
}

export interface AgentRunRequest {
  agent_type: string
  model_config: ModelConfig
  prompt: string
  data: Record<string, any>
  tables: Record<string, any[]>
  session_id: string
  config?: Record<string, any>
}

export interface AgentFollowupRequest extends AgentRunRequest {
  previous_dialog: any[]
}

export interface AgentResponse {
  status: 'ok' | 'error' | 'other error'
  content: Record<string, any>
  code?: string
  dialog: any[]
  agent: string
  refined_goal?: Record<string, any>
  error_message?: string
  execution_time?: number
}

export interface ModelListResponse {
  models: Array<{
    provider: string
    model: string
    name: string
  }>
  providers: string[]
}

export class AgentAPI {
  /**
   * Run an AI agent for data transformation
   * Migrated from Flask /api/agent/run endpoint
   */
  static async runAgent(request: AgentRunRequest): Promise<AgentResponse> {
    try {
      return await apiClient.post<AgentResponse>('/api/v1/agents/run', request)
    } catch (error) {
      if (error instanceof APIError) {
        throw error
      }
      throw new APIError('Failed to run agent', 500, error)
    }
  }

  /**
   * Follow up with an AI agent for iterative refinement
   */
  static async followupAgent(request: AgentFollowupRequest): Promise<AgentResponse> {
    try {
      return await apiClient.post<AgentResponse>('/api/v1/agents/followup', request)
    } catch (error) {
      if (error instanceof APIError) {
        throw error
      }
      throw new APIError('Failed to followup with agent', 500, error)
    }
  }

  /**
   * Get list of available LLM models and providers
   */
  static async getAvailableModels(): Promise<ModelListResponse> {
    try {
      return await apiClient.get<ModelListResponse>('/api/v1/agents/models')
    } catch (error) {
      if (error instanceof APIError) {
        throw error
      }
      throw new APIError('Failed to get available models', 500, error)
    }
  }

  /**
   * Get agent configuration settings
   */
  static async getAgentConfig(): Promise<Record<string, any>> {
    try {
      return await apiClient.get<Record<string, any>>('/api/v1/agents/config')
    } catch (error) {
      if (error instanceof APIError) {
        throw error
      }
      throw new APIError('Failed to get agent config', 500, error)
    }
  }

  /**
   * Explain generated code in natural language
   */
  static async explainCode(
    code: string,
    modelConfig: ModelConfig,
    sessionId: string
  ): Promise<AgentResponse> {
    try {
      return await apiClient.post<AgentResponse>('/api/v1/agents/explain-code', {
        code,
        model_config: modelConfig,
        session_id: sessionId,
      })
    } catch (error) {
      if (error instanceof APIError) {
        throw error
      }
      throw new APIError('Failed to explain code', 500, error)
    }
  }
}

// Example usage function for the /api/agent/run endpoint
export async function runDataTransformationAgent(
  prompt: string,
  data: Record<string, any>,
  tables: Record<string, any[]>,
  modelConfig: ModelConfig,
  sessionId: string
): Promise<AgentResponse> {
  const request: AgentRunRequest = {
    agent_type: 'python_data_transform',
    model_config: modelConfig,
    prompt,
    data,
    tables,
    session_id: sessionId,
  }

  return AgentAPI.runAgent(request)
}

// Hook-style API for React components
export function useAgentAPI() {
  return {
    runAgent: AgentAPI.runAgent,
    followupAgent: AgentAPI.followupAgent,
    getAvailableModels: AgentAPI.getAvailableModels,
    getAgentConfig: AgentAPI.getAgentConfig,
    explainCode: AgentAPI.explainCode,
  }
}